### PR TITLE
Test Delta Lake metadata handling when schema already exists

### PR DESCRIPTION
**Pull request by pajaks**
## Description
Add test for https://github.com/trinodb/trino/pull/13242 which was merged without a test as we didn't have an infrastructure for such tests at that time.
## Additional context and related issues
https://github.com/trinodb/trino/issues/15282

## Release notes

(x) This is not user-visible or docs only and no release notes are required.

### CODE REVIEW

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/metastore/glue/TestDeltaLakeGlueMetastore.java [hunk 1]
```diff
@@ -241,6 +252,37 @@ public void testHideNonDeltaLakeTable()
...
+
+        metadata.createSchema(session, testSchemaName, Map.of(), TRINO_PRINCIPAL);
+        metadata.createSchema(session, testSchemaName, Map.of(), TRINO_PRINCIPAL);
+
+        // Clean up schema table
+        metastoreClient.dropDatabase(testSchemaName, true);
+    }
+
+    @Test
+    public void testFailWhenCreateTheSameSchemaFromTwoSessions()
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/metastore/glue/TestDeltaLakeGlueMetastore.java:270]: `testFailWhenCreateTheSameSchemaFromTwoSessions` -> `testCreateSameSchemaFromDifferentSessionsFails`

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/metastore/glue/TestDeltaLakeGlueMetastore.java [hunk 2]
```diff
@@ -241,6 +252,37 @@ public void testHideNonDeltaLakeTable()
...
 
+    @Test
+    public void testDoubleCreateSchemaForTheSameSession()
+    {
+        String testSchemaName = "test_schema_" + randomName();
+
+        DeltaLakeMetadata metadata = metadataFactory.create(session.getIdentity());
+
+        metadata.createSchema(session, testSchemaName, Map.of(), TRINO_PRINCIPAL);
+        metadata.createSchema(session, testSchemaName, Map.of(), TRINO_PRINCIPAL);
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/metastore/glue/TestDeltaLakeGlueMetastore.java:263]: Maybe call `metadata.schemaExists()` after the second call to be clear that the schema exists before doing the second call.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/metastore/glue/TestDeltaLakeGlueMetastore.java [hunk 3]
```diff
@@ -241,6 +252,37 @@ public void testHideNonDeltaLakeTable()
                 .isEmpty();
     }
 
+    @Test
+    public void testDoubleCreateSchemaForTheSameSession()
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/metastore/glue/TestDeltaLakeGlueMetastore.java:256]: `testDoubleCreateSchemaForTheSameSession` -> `testCreateSchemaTwiceInTheSameSession`

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/metastore/glue/TestDeltaLakeGlueMetastore.java [hunk 4]
```diff
@@ -241,6 +252,37 @@ public void testHideNonDeltaLakeTable()
...
+    {
+        String testSchemaName = "test_schema_" + randomName();
+
+        DeltaLakeMetadata metadata = metadataFactory.create(session.getIdentity());
+
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/metastore/glue/TestDeltaLakeGlueMetastore.java:261]: try/finally to ensure that the Glue database does get deleted even in case of exceptions

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/metastore/glue/TestDeltaLakeGlueMetastore.java (hunk 2 above)
Comment (homar) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/metastore/glue/TestDeltaLakeGlueMetastore.java:263]: In this and the second test case shouldn't we validate that schema was actually created?

**Comment by findinpath** [comment]
Commit message `Add tests for creating schema` 

Maybe `Test Delta AWS Glue handling when schema already exists`

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/metastore/glue/TestDeltaLakeGlueMetastore.java (hunk 2 above)
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/metastore/glue/TestDeltaLakeGlueMetastore.java:263]: Just calling `DeltaLakeMetadata#createSchema` doesn't trigger the issue in #15177. The situation is remote metastore succeeded, but the client got timed out in **_a_** createSchema operation.

**Review by ebyhr — COMMENTED** [review]
~The new test `testDoubleCreateSchemaForTheSameSession` should fail without commits in https://github.com/trinodb/trino/pull/15177. Currently, it passes. I would recommend reverting commits temporarily when writing the test.~ I realized the issue number in the description was wrong.

**Comment by pajaks** [comment]
> 

It was mistake in description. Tests are for [#13242](https://github.com/trinodb/trino/pull/13242)

**Comment by ebyhr** [comment]
> Tests are for https://github.com/trinodb/trino/pull/13242

The PR isn't specific to Glue. Could you explain why this PR add tests only for Glue?

**Comment by pajaks** [comment]
> > Tests are for #13242
> 
> The PR isn't specific to Glue. Could you explain why this PR add tests only for Glue?

Test was in wrong place. I moved it to generic tests and changed title.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeMetadata.java [hunk 5]
```diff
@@ -465,6 +473,35 @@ public void testGetInputInfoForUnPartitionedTable()
         assertThat(deltaLakeMetadata.getInfo(tableHandle)).isEqualTo(Optional.of(new DeltaLakeInputInfo(false)));
     }
 
+    @Test
+    public void testCreateSchemaTwiceInTheSameSession()
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeMetadata.java:477]: This feels like a unit test and somebody without a context for the `SchemaAlreadyExistsException` problem would have a hard time to figure out what are the tests there.
What about using a specific test - similar to `TestIcebergFileMetastoreTableOperationsInsertFailure` ?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+import static com.google.common.io.MoreFiles.deleteRecursively;
+import static com.google.common.io.RecursiveDeleteOption.ALLOW_INSECURE;
+import static com.google.inject.util.Modules.EMPTY_MODULE;
+import static io.trino.plugin.deltalake.DeltaLakeConnectorFactory.CONNECTOR_NAME;
+import static io.trino.plugin.hive.HiveTestUtils.HDFS_ENVIRONMENT;
+import static io.trino.testing.TestingNames.randomNameSuffix;
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static java.lang.String.format;
+
+public class TestDeltaLakeSchemaCreate
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java:45]: How about `TestDeltaLakeCreateSchemaInternalRetry` or something?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                new HiveMetastoreConfig().isHideDeltaLakeTables(),
+                new FileHiveMetastoreConfig()
+                        .setCatalogDirectory(dataDirectory)
+                        .setMetastoreUser("test"))
+        {
+            @Override
+            public synchronized void createDatabase(Database database)
+            {
+                super.createDatabase(database);
+                afterCreateDatabaseFunction.run();
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java:78]: I'm not sure why `Runnable` is needed. I would throw an exception directly here.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+            deleteRecursively(Path.of(dataDirectory), ALLOW_INSECURE);
+        }
+    }
+
+    @Test
+    public void testSchemaCreationWithTimeout()
+    {
+        // Simulate retry mechanism with timeout failure.
+        // 1. createDatabase correctly create schema but timeout is triggered
+        // 2. Retry to createDatabase throws SchemaAlreadyExistsException
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java:102]: Move this comment to before `super.createDatabase(database);` if you inline `Runnable`.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        afterCreateDatabaseFunction = () -> {
+            throw new SchemaAlreadyExistsException("Test-simulated metastore schema already exist exception");
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java:104]: The argument takes a schema name, not an error message.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        };
+
+        assertQuerySucceeds(format("CREATE SCHEMA %s", TEST_SCHEMA));
+        assertQuery(format("SHOW SCHEMAS LIKE '%s'", TEST_SCHEMA), format("VALUES ('%s')", TEST_SCHEMA));
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java:108]: String concatenation provides SQL highlight. We use this style when the query is simple.
```java
        assertQuerySucceeds("CREATE SCHEMA " + TEST_SCHEMA);
        assertQuery("SHOW SCHEMAS LIKE '" + TEST_SCHEMA + "'", "VALUES '" + TEST_SCHEMA + "'");
```

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    }
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java:109]: Optional: Ideally, we should add another test method to verify that `CREATE SCHEMA` throws an exception when `presto_query_id` is different from the session query id.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeMetadata.java [hunk 12]
```diff
@@ -465,6 +473,35 @@ public void testGetInputInfoForUnPartitionedTable()
...
+
+        deltaLakeMetadata.createSchema(SESSION, testSchemaName, Map.of(), TRINO_PRINCIPAL);
+        assertThat(deltaLakeMetadata.schemaExists(SESSION, testSchemaName)).isTrue();
+
+        ConnectorSession secondSession = TestingConnectorSession.builder().build();
+        assertThatThrownBy(() -> deltaLakeMetadata.createSchema(secondSession, testSchemaName, Map.of(), TRINO_PRINCIPAL))
+                .isInstanceOf(SchemaAlreadyExistsException.class)
+                .hasMessage(format("Schema already exists: '%s'", testSchemaName));
+        assertThat(deltaLakeMetadata.schemaExists(SESSION, testSchemaName)).isTrue();
+    }
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeMetadata.java:503]: Can we revert this change in this class or add some comments for clarifying the test purpose?

**Review by ebyhr — COMMENTED** [review]
Almost good to me. Thanks for adding this test.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java (hunk 7 above)
Comment (pajaks) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java:78]: My idea was to make test itself more self describing, but I agree that it my be unnecessary complication.

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreateInternalRetry.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+import static com.google.common.io.RecursiveDeleteOption.ALLOW_INSECURE;
+import static com.google.inject.util.Modules.EMPTY_MODULE;
+import static io.trino.plugin.deltalake.DeltaLakeConnectorFactory.CONNECTOR_NAME;
+import static io.trino.plugin.hive.HiveMetadata.PRESTO_QUERY_ID_NAME;
+import static io.trino.plugin.hive.HiveTestUtils.HDFS_ENVIRONMENT;
+import static io.trino.testing.TestingNames.randomNameSuffix;
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
+
+public class TestDeltaLakeSchemaCreateInternalRetry
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreateInternalRetry.java:48]:
```suggestion
public class TestDeltaLakeCreateSchemaInternalRetry
```

**Comment by ebyhr** [comment]
> Add tests for creating schema

@pajaks Could you fix the commit title as "Add tests for internal retry to create schema in Delta Lake"?

**Review by findinpath — APPROVED** [review]

**Comment by ebyhr** [comment]
@cla-bot check

**Comment by ebyhr** [comment]
@pajaks Are you going to add tests for Hive as well?

**Comment by pajaks** [comment]
> @pajaks Are you going to add tests for Hive as well?

It seems like a good idea. I will look into it.
